src/Collections/Arrays/Arrays.dfy

```diff
@@ -0,0 +1,374 @@
+// RUN: %dafny /compile:0 "%s"
+
+/*******************************************************************************
+*  Copyright by the contributors to the Dafny Project
+*  SPDX-License-Identifier: MIT 
+*******************************************************************************/
+
+include "../Sets/Sets.dfy"
+include "../../Comparison.dfy"
+
+module Arrays{
+
+import opened Sets
+import opened Comparison
+import opened Wrappers
+
+  trait Sorter<T> {
```

Comment:
This definitely needs an example in `examples` to demonstrate using it. I assume you have to not only provide a concrete class that implements `Sorter<T>` and fills in `Cmp`, but perhaps even has to invoke a lemma or two from the trait in order to prove that it works.

Is it possible to also provide a `PredicateSorter` class that is constructed around a `(T, T) -> Cmp` function value? Perhaps it could even be a partial function if you also provide a ghost set of all values you intend to sort with it.

Reply:
Just to clarify, I'd strongly prefer to have an example, but the `PredicateSorter` idea can wait for a future PR.

Reply:
Oh yikes, I went to write an example just to play around with it and immediately got the `A class may only extend a trait in the same module, unless the parent trait is annotated with {:termination false}` error. That means this is currently useless unless we add `{:termination false}`.

Unfortunately we should treat that as a hard blocker given there's already been resistance to putting `{:termination false}` on any standard library code: https://github.com/dafny-lang/libraries/pull/22 And especially given the discovery of https://github.com/dafny-lang/dafny/issues/2500 I'm not inclined to push back on that.

We can think about whether there's a good alternative for this code that doesn't use traits, but it might be better to stick with this if it's good UX and at least wait for the solution to https://github.com/dafny-lang/dafny/issues/2500.

Reply:
That's not right: the proper API to use is `PredicateSorter`, which is a regular class, defined right below `Sorter`.  That doesn't require `:termination false`.

Original code here, with examples:  https://github.com/dafny-lang/compiler-bootstrap/blob/4f616822209828e48cf63d3da66ee1c010f689d4/src/Utils/Lib.Sort.dfy#L396-L420